node/src/components/storage.rs

```diff
@@ -1578,24 +1530,33 @@ fn initialize_block_body_v1_db(
 ) -> Result<(), LmdbExtError> {
     info!("initializing v1 block body database");
     let mut txn = env.begin_rw_txn()?;
+
+    let block_body_hash_to_header_map =
+        construct_block_body_to_block_header_reverse_lookup(&txn, block_header_db)?;
+
     let mut cursor = txn.open_rw_cursor(*block_body_v1_db)?;
 
     for (raw_key, _raw_val) in cursor.iter() {
-        if deleted_block_body_hashes_raw.contains(raw_key) {
+        let key =
+            Digest::try_from(raw_key).map_err(|err| LmdbExtError::DataCorrupted(Box::new(err)))?;
+        if !block_body_hash_to_header_map.contains_key(&key) {
+            if !deleted_block_body_hashes_raw.contains(raw_key) {
+                // This means that the block body isn't referenced by any header, but no header
+                // referencing it was just deleted, either
```

Comment:
> but no header referencing it was just deleted, either

`deleted_block_body_hashes` should contain block bodies, not headers.

Reply:
It _does_ contain the block body hashes - the ones that were included in blocks that were deleted. We used to just delete all bodies listed here, but that's incorrect (see https://github.com/casper-network/casper-node/pull/1649#discussion_r662132791) - now I'm deleting bodies that have no corresponding header, and this part only checks whether, if we have such a body, it was included in a block that was just deleted. If it wasn't, it indicates that we had an entry in the block body DB that wasn't a part of any block even before we deleted any blocks, and that could indicate a bug.

Reply:
I agree about deleting unreachable elements :+1: